common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+        }
+
+        def trySet(expectedState: State, desiredState: State): Boolean = {
+            state.compareAndSet(expectedState, desiredState)
+        }
+    }
+
+    protected def cacheInvalidate(key: Any)(
+        implicit transid: TransactionId, logger: Logging): Unit = {
+        if (cacheEnabled) {
+            logger.info(this, s"invalidating $key")
+            cache remove key
+        }
+    }
+
+    protected def cacheLookup[Wsuper >: W](
+        key: Any,
+        future: => Future[W],
+        fromCache: Boolean = cacheEnabled)(
+            implicit ec: ExecutionContext, transid: TransactionId, logger: Logging): Future[W] = {
+
+        if (fromCache) {
+            val promise = Promise[W]
+            val desiredEntry = Entry(transid, ReadInProgress, promise.future)
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) flatMap { actualEntry =>
+                // ... and see what we get back
+
+                actualEntry.state.get match {
+                    case Cached => {
+                        logger.debug(this, s"Cached read $key")
+                        makeNoteOfCacheHit(key)
+                        actualEntry.value
+                    }
+
+                    case ReadInProgress => {
+                        if (actualEntry.transid == transid) {
+                            logger.debug(this, "Read initiated");
+                            makeNoteOfCacheMiss(key)
+                            listenForReadDone(key, actualEntry, future, promise)
+                            actualEntry.value
+
+                        } else {
+                            logger.debug(this, "Coalesced read")
+                            makeNoteOfCacheHit(key)
+                            actualEntry.value
+                        }
+                    }
+
+                    case WriteInProgress | InvalidateInProgress => {
+                        logger.debug(this, "Reading around a write in progress")
+                        makeNoteOfCacheMiss(key)
+                        future
+                    }
+                }
+            }
+
+        } else {
+            // not caching
+            future
+        }
+    }
+
+    protected def cacheUpdate(doc: W, key: Any, future: Future[Winfo])(
+        implicit transid: TransactionId, logger: Logging, ec: ExecutionContext): Future[Winfo] = {
+
+        if (cacheEnabled) {
+            logger.info(this, s"caching $key")
+
+            val desiredEntry = Entry(transid, WriteInProgress, Future { doc })
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) map { actualEntry =>
+                // ... and see what we get back
+
+                if (transid == actualEntry.transid) {
+                    //
+                    // then we won the race, and now own the entry
+                    //
```

Comment:
Would you remove the "empty" comment lines? I feel like they're stretching the code, fitting less on the screen while reading through.